### ISSUE TITLE
Fetching current input parameters values (for a service or a binding)

### ISSUE THREAD

**Issue by mattmcneeney** [openservicebrokerapi/servicebroker#159]
Following on from [Issue 59 (Service broker can declare supported schema for parameters field)](https://github.com/openservicebrokerapi/servicebroker/issues/59), we are expecting that platforms will want to be able to fetch the current values of any service or binding input parameters.

How could this be integrated into the service broker API?

---

**Potential idea**

Add two new endpoints:
- `GET /v2/service_instances/:instance_id` to fetch input parameters for a service.
- `GET /v2/service_instances/:instance_id/service_bindings/:binding_id` to fetch input parameters for a binding.

Concerns:
- Security
- Service brokers having to maintain some kind of state

---

### Proposal Doc 
(edit by @avade)

[Proposal: Fetching Configuration Parameters](https://docs.google.com/document/d/13EviTJXm9-P-SBqeGexyEy_wclKPCr1S3ZGVDpaC0Tw/edit?usp=sharing)

**Comment by angarg12** [comment]
This seems to come back to the bigger question of whether OSB should provide a full REST API. Also I think that the common feeling around here is not to require brokers to keep state.

If we have specific use cases where clients require to retrieve the value of parameters it would make for a much compelling argument in favour of this requirement, instead of guessing a possible future need.

In any case it is probably better to raise this question in the weekly meeting.

**Comment by mattmcneeney** [comment]
Of course, I've added it to the agenda for today's call. There are some use cases within Cloud Foundry whereby app developers would want to retrieve the configuration of their service instances and bindings. It will be interesting to hear if anyone else has run into this problem.

**Comment by gberche-orange** [comment]
@mattmcneeney 

> Following on from Issue 59 (Service broker can declare supported schema for parameters field), we are expecting that platforms will want to be able to fetch the current values of any service or binding input parameters.
> How could this be integrated into the service broker API?

SB receive service instance and service binding parameters as in the ``parameters`` request field of the https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#provisioning request and would perform validation of received params against the schema declared in the catalog. Platform may optionally also validate the parameters field as they flow them to brokers.

I'm not sure to understand why platforms would need to fetch the current values of parameters of any previously provisioned service instance or binding. Can you please detail the rationale behind this ?

**Comment by gberche-orange** [comment]
Use case mentionned during the WG call is that client UIs could display previously submitted parameters fields as to improve “update service instance” UX.

Would be worth investigating whether platforms could expose these endpoints instead of brokers, keeping the OSB API simple and reducing state management constraints on brokers.

In order to avoid allow clients UI fetching and displaying sensitive params:
1- brokers could declare some fields as being sensitive in the broker schemas so that UIs can expect not to fetch & display these fields. 
2- alternatively, if the persistence and serving of the field does remain the responsibility of brokers, brokers could choose to not implement these endpoints or partially implementing them (e.g. not returning sensitive fields).

**Comment by mattmcneeney** [comment]
As well as improving the update service instance UX, this could also improve visibility into existing service instances and bindings. CF is working on a number of service brokers that expose non-sensitive parameters (such as number of nodes in a cluster, disk sizes, etc) that developers would like to be able to see and change via a CLI or UI.

We have already come across use cases whereby brokers have their own dashboards for modifying configuration, so the platform will not always know the correct parameter values, so relying on the platform to provide current parameter values will in some cases be impossible.

If we made these API endpoints optional, that would mean that brokers that are unable to maintain state do not have to, and we can introduce this in version 2 (non-breaking change). Note that we also discussed on the call how some brokers could rely on other services to maintain state.

We could also specify how brokers (if they want to) should respond when asked for parameters that contain sensitive information. For example, a call to:
```
GET /v2/service_instances/:instance_id
```
could return
```
{
  "parameters": {
    "nodes": 10,
    "password": "<redacted>"
}
```
Alternatively brokers could simply not return sensitive parameters. I'm unsure whether we should specify how brokers should handle secrets within the spec.

@gberche-orange What are your thoughts on this?
@pmorie Have you come across any kubernetes use cases for this?

As CF is now beginning work on the input parameter schemas for the validation through implementation, we're happy to do the same for this piece of work if we can agree on a sensible approach.

**Comment by mattmcneeney** [comment]
I've put together a document that specifies the Cloud Foundry use case for this, and a proposed solution. Thoughts and feedback very welcome.
[Proposal: Fetching Configuration Parameters](https://docs.google.com/document/d/13EviTJXm9-P-SBqeGexyEy_wclKPCr1S3ZGVDpaC0Tw/edit?usp=sharing)

**Comment by mcelep** [comment]
I've had this request on CC a while back: https://github.com/cloudfoundry/cloud_controller_ng/issues/680

**Comment by duglin** [comment]
Ignoring the security aspects of this, I like the idea of supporting GET - we'd just need to make it optional.

**Comment by youngm** [comment]
Today all of our service brokers expose an INFO end point that can be queried to get information about a service instance.  We don't supply sensitive data and we display the info returned in our Web UI.  

It would be nice if the CC could invoke a GET endpoint on the broker where the broker could provide information about a service instance or binding.  I would love for support for this to be more formal.

**Comment by mattmcneeney** [comment]
Hey all. So I've started mocking this together on a branch. My main question is:
- Should this change _only_ add the GET endpoints returning the `parameters` object, or should the endpoints also return the service instance / binding data returned after a provision / binding creation?

Let me know what you think of [https://github.com/openservicebrokerapi/servicebroker/compare/master...mattmcneeney:get-endpoints](https://github.com/openservicebrokerapi/servicebroker/compare/master...mattmcneeney:get-endpoints).

Cheers.

**Comment by nilebox** [comment]
@mattmcneeney considering that even a single GET request makes service broker _stateful_, I would add GET endpoints for all corresponding PUT requests.

**Comment by duglin** [comment]
From the f2f notes I see:
```
Add a GET endpoint for service instances
Add a GET endpoint for service binds 
```
My take-away from the f2f was that we would add a GET endpoint for all resources.

**Comment by mattmcneeney** [comment]
I think the branch has both of those GET endpoints.

**Comment by mattmcneeney** [comment]
Any further thoughts on this all?

**Comment by shalako** [comment]
> Should this change only add the GET endpoints returning the parameters object, or should the endpoints also return the service instance / binding data returned after a provision / binding creation?

I recommend supporting all fields currently supported for responses to the create instance or create binding endpoints, as suggested in the diff linked above.

This proposal looks good to me and I would recommend for moving it to validation-via-implementation phase.

**Comment by mattmcneeney** [comment]
Sounds good to me @shalako. @pmorie @duglin and other Kubernetes folks; are you happy to start validating the implementation of this?

**Comment by duglin** [comment]
I think you can remove text like:
```
All response bodies MUST be a valid JSON Object ({}). This is for future compatibility;
```

**Comment by mattmcneeney** [comment]
@duglin I've removed the following text from my PR:
```
All response bodies MUST be a valid JSON Object (`{}`). This is for future compatibility; it will be easier to add fields in the future if JSON is expected rather than to support the cases when a JSON body might or might not be returned.
```

As discussed on the call, we probably want this to be removed in the existing spec :)

**Comment by svrc-pivotal** [comment]
Just adding a note here that one other use cases of requiring parameter access is to be able to do an unbind/rebind operation on a service and the operator needs to know how the service was initially bound (this is not likely the person who did the binding in the first place!).   

This is a useful operation to do if the backing service has changed hostnames or IPs, such as in a disaster recovery scenario.

**Comment by nilebox** [comment]
> This is a useful operation to do if the backing service has changed hostnames or IPs, such as in a disaster recovery scenario.

+1 to this.
The similar use case we think of is credentials rotation on the broker side.

**Comment by shalako** [comment]
@mattmcneeney @duglin 
Re: removing `All response bodies MUST be a valid JSON Object ({}). This is for future compatibility;`

CF will consider a broker response invalid if it does not include a valid JSON object.

**Comment by duglin** [comment]
@shalako right but since we define a Body I think that sentence is redundant. I think it was only needed before in cases where there was no Body properties defined.

**Comment by mattmcneeney** [comment]
Can we just add that in one place to the top of the spec and remove from all existing response bodies?

**Comment by mattmcneeney** [comment]
Branch updated: https://github.com/mattmcneeney/servicebroker/tree/get-endpoints

**Comment by shalako** [comment]
I see this has been promoted to validation. Great news.

**Comment by avade** [comment]
@mattmcneeney Do you expect this to be validated in CF and PR'd in time for 2.13 (early September)?

@pmorie will the service-catalog SIG be implementing this feature soon?

**Comment by mattmcneeney** [comment]
@avade I don't think we'll have feedback on this one by early September unfortunately. We may have support for it in the platform by then, but we probably won't have any brokers supporting the feature to get feedback from.

**Comment by mattmcneeney** [comment]
@duglin The current branch ( https://github.com/mattmcneeney/servicebroker/blob/get-endpoints/spec.md#fetching-a-service-instance ) only has a `200` response code to the new GET endpoints.

All - do we need to add a `202 Accepted` code for when an async operation (currently only for service instance provisioning) is in progress but the resource is not yet ready? Or can the last operation endpoint be relied on for this (when that returns `"state": "succeeded", the GET endpoint should be immediately available)?

**Comment by duglin** [comment]
Couple of comments:
- on instances you don't say that the `parameters` should match the incoming `parameters` from the provision request. While its true that the broker might actually change the values for some reason, it might be good to at least correlate the two because as it stands now the text doesn't imply there's any relationship between them at all. I'm hoping that in most cases they will be the exact same otherwise its a bit misleading.
- it talks about matching the schema - does that mean that the schema PR must be merged first?
- fetching a binding should probably be in the Binding section - and fetching an instance should be in the provisioning section - actually, would it make sense to not have the TOC list these separately?  Should we instance just have a "Managing Instances" type of section and talk about the CRUD ops for Instances - and then do the same for Bindings?  Otherwise to get all of the info for dealing with each you need to jump around a lot.
- I wonder if it should return the other info, like context, service-id, etc...

In the binding:
- s/as a failure database/as a failure/
- should we include the other info - like app_guid, bind_resource, ... ?

The question about 202 doesn't just apply during async, but even sync since a GET could be issued while a provision/bind request is being worked on in a synchronous model.

Overall it seems right, we just need to decide on which fields to return.

Didn't we also talk about allowing a broker to return a subset of the data if it didn't want (or couldn't) return all of it - like the creds?

**Comment by Samze** [comment]
The platform currently has a way to figure out which optional endpoints the broker supports.

1. Update - `plan_updateable: true` in the catalog
1. Bind/Unbind - `bindable: true` in the catalog
1. Last Operation - Provision returns 202 instead of 200

This allows platforms to only send requests to brokers who explicitly declare support for them. The proposal as it stands does not allow any discoverability mechanism for the platform.

I think in the face to face we discussed that brokers who do not support this endpoint can just return a 404, however I see a number of flaws with this approach:
1. The platform has to make a request to the broker each time to find out if it supports the endpoint. This could be cached but I'm not sure how you would invalidate this.
1. The platform cannot distinguish between a broker that does not support this endpoint or a broker that temporarily is failing. e.g. routes becoming unregistered and responds with 404s.
1. Old brokers may have implemented that route for some other purpose outside of the spec.

Instead should we follow current convention and add explicit support in the catalog for these endpoints?

**Comment by skaegi** [comment]
Really late to the game ... my apologies.
Going back to the original use-case, we've been supporting this by having the broker return back a "parameters" property (similar to "credentials") in the response object for provisioning/binding creates and updates. 

The idea here is that it is up to the broker to decide which parameters it wants the controller to save. In our implementations fields like passwords, authentication keys, as well as one-time configuration values are not typically returned in this set of parameters. If an admin elects to modify a service instance or binding these saved parameters are used to populate "defaults" in our UI.

--
I'm not saying GET isn't necessarily a good idea, just that there are other approaches to consider that won't add a new endpoint.

**Comment by mattmcneeney** [comment]
@skaegi We are seeing many brokers use platform components or the data service it backs onto to store any state it has to keep. I agree though that it is entirely up to the brokers what they want to return in these endpoints, from parameter values to credentials and anything else.

**Comment by skaegi** [comment]
re: platform components -- that's interesting!
Do you have a mechanism to tell the brokers how they might access a storage service or something similar as part of the registration flow.

**Comment by mattmcneeney** [comment]
Closing this as it will be resolved by #333
